### PR TITLE
Adding LiDAR Feature

### PR DESCRIPTION
**Pull request by Amirsaeid254**
This pull request introduces the LiDAR feature. It includes a configuration file for adjusting LiDAR settings and a main LiDAR class compatible with both normal and vectorized engines.

The core functionality involves creating a polar mesh based on the LiDAR configuration and extracting the map boundaries within the LiDAR range.

### CODE REVIEW

File: obstacle_rich_env/envs/robot.py [hunk 1]
```diff
@@ -17,6 +19,9 @@ def make_dynamics(self):
             return dynamics_class(params=self._params, random_generator=self.random_generator)
         raise 'Robot dynamics not defined'
 
+    def mount_lidar(self, map, config):
+        self.lidar = Lidar(map, self, config)
```
Comment (pedramrabiee) [inline obstacle_rich_env/envs/robot.py]: Please move lidar instantiation to Engine

File: obstacle_rich_env/configs/lidar_configs.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from attrdict import AttrDict as AD
+import torch
+
+lidar_configs = AD(max_range=5,
+                   ray_num=100,
+                   scan_angle=[-torch.pi, torch.pi],
+                   update_rate=0.2,
+                   ray_sampling_rate=1000,
+                   space_dimension=2,
+                   sensor_mounting_location=[0, 0],
```
Comment (pedramrabiee) [inline obstacle_rich_env/configs/lidar_configs.py]: Can we do XYZ, RPY instead? or if you want to keep it 2D, change it to:
sensor_mouting_xy
sensor_mounting_yaw

File: obstacle_rich_env/envs/lidar.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        c, s = torch.cos(self.sensor_angle), torch.sin(self.sensor_angle)
+
+        x = c * points[..., 0] - s * points[..., 1] + self.sensor_location[0]
+        y = s * points[..., 0] + c * points[..., 1] + self.sensor_location[1]
+
+        return torch.stack((x, y), dim=-1)
+
+    def _boundary_extractor(self, mesh):
+        # Calculate barrier values
+        z = self.map.nonsmoothcomp_pos_barrier.min_barrier(mesh)
```
Comment (pedramrabiee) [inline obstacle_rich_env/envs/lidar.py]: Can you report the share of the computational cost of computing the barrier to that of _boundary_extractor?

File: obstacle_rich_env/configs/common_configs.py [hunk 4]
```diff
@@ -7,7 +7,8 @@
     max_episode_steps=100,
     observation_flatten=False,
     # obs_key_to_return=["state", "custom_state", "goal_robot_diff", "barriers", "min_barrier"],        # provide a list of keys to include in the observation
-    obs_key_to_return=["state", "custom_state", "goal_robot_diff", "barriers"],          # provide a list of keys to include in the observation
+    # obs_key_to_return=["state", "custom_state", "goal_robot_diff", "barriers"],          # provide a list of keys to include in the observation
+    obs_key_to_return=["state", "custom_state", "goal_robot_diff", "barriers", 'obs_lidar'],
```
Comment (pedramrabiee) [inline obstacle_rich_env/configs/common_configs.py]: I suggest changing the key name to obstacle_lidar

File: obstacle_rich_env/envs/robot.py (hunk 1 above)
Comment (Amirsaeid254) [reply obstacle_rich_env/envs/robot.py]: Lidar instantiation moved to the engine.

File: obstacle_rich_env/configs/lidar_configs.py (hunk 2 above)
Comment (Amirsaeid254) [reply obstacle_rich_env/configs/lidar_configs.py]: For this version, I would like to keep it as 2D. In the future, I will update the LiDAR class to a more general version to cover both 3D and 2D environments. Variable names have been changed as you suggested.

File: obstacle_rich_env/configs/common_configs.py (hunk 4 above)
Comment (Amirsaeid254) [reply obstacle_rich_env/configs/common_configs.py]: Variable names have been changed as you suggested.

**Pull request merged**